File: btypes_primitives/src/main/cpp/BBoolean.cpp
#include <iostream>
#include <string>
#include "BBoolean.hpp"


BBoolean::BBoolean(bool val) {
    value = val;
}

BBoolean::BBoolean(){};

BBoolean::BBoolean(const BBoolean& val) {
    value = val.value;
}

bool BBoolean::booleanValue() const {
    return value;
}

BBoolean BBoolean::_or(const BBoolean& other) const {
    return value || other.value;
}

BBoolean BBoolean::_or(bool other) const {
    return value || other;
}

BBoolean BBoolean::_xor(const BBoolean& other) const {
    return value ^ other.value;
}

BBoolean BBoolean::_xor(bool other) const {
    return value ^ other;
}

BBoolean BBoolean::_and(const BBoolean& other) const {
    return value && other.value;
}

BBoolean BBoolean::_and(bool other) const {
    return value && other;
}

BBoolean BBoolean::_not() const {
    return !value;
}

BBoolean BBoolean::implies(const BBoolean& other) const {
    return !value || other.value;
}

BBoolean BBoolean::implies(bool other) const {
    return !value || other;
}

BBoolean BBoolean::equivalent(bool other) const {
    return value == other;
}

BBoolean BBoolean::equivalent(const BBoolean& other) const {
    return value == other.value;
}

BBoolean BBoolean::equal(const BBoolean& other) const {
    return value == other.value;
}

BBoolean BBoolean::unequal(const BBoolean& other) const {
    return value != other.value;
}

bool BBoolean::operator <(const BBoolean& other) {
    return value != other.value;
}

void BBoolean::operator =(const BBoolean& other) {
    value = other.value;
}

int BBoolean::hashCode() const {
    return value == true ? 1 : 0;
}

File: benchmarks/execution/JS/btypes/BBoolean.ts
import {BObject} from "./BObject.js";

export class BBoolean implements BObject{
    readonly value: boolean;

    constructor(value: boolean | string | BBoolean) {
        if (typeof (value) === 'boolean') {
            this.value = value;
        } else if (typeof value === 'string') {
            this.value = value.toLowerCase() === 'true';
        } else {
            this.value = value.value;
        }
    }
    
    and(other: BBoolean): BBoolean {
        return new BBoolean(this.value && other.value)
    }

    or(other: BBoolean): BBoolean {
        return new BBoolean(this.value || other.value)
    }

    not(): BBoolean {
        return new BBoolean(!this.value)
    }

    implies(other: BBoolean): BBoolean {
        return this.not().or(other)
    }

    equivalent(other: BBoolean): BBoolean {
        return this.implies(other).and(other.implies(this))
    }

    equal(other: BBoolean): BBoolean {
        return new BBoolean(this.value === other.value);
    }

    equals(other: any): boolean {
        if (!(other instanceof BBoolean)) {
            return false;
        }
        return this.equal(other).booleanValue();
    }

    unequal(other: BBoolean): BBoolean {
        return new BBoolean(this.value != other.value);
    }

    booleanValue(): boolean {
        return this.value;
    }

    toString(): string {
        return this.value.toString();
    }

    hashCode(): number {
        return (31 * 1) ^ (this.value == true ? 1 : 0 << 1);
    }

}
